### ISSUE TITLE
Parameter Size Display Bug

### ISSUE THREAD

**Issue by TheoAC** [rust-lang/rust-analyzer#19517]
I've found what I understand to be a minor bug in how rust-analyzer displays size information for variables. In a generic function, parameters of type  `&T` are assumed to be one word in size, even with the type bound `T: ?Sized`. But if `T` happens to be a dynamically sized type, then a parameter of type `&T` will be a wide pointer, which is two words in size. In this case, we have an inconsistency:
```rust
// parameter `_x` has `size = 8, align = 0x8`
fn foo<T: ?Sized>(_x: &T) {}

fn main() {
    // variable `x` has `size = 16 (0x10), align = 0x8`
    let x = "";
    foo(x);
}
```
The comments in the code above refer to what rust-analyzer displays when hovering the mouse over the declaration of the parameter/variable. The issue here is that the display for `_x` as a parameter of `foo` is not necessarily correct.

**rust-analyzer version**:
0.3.2362-standalone (fb133c8c80 2025-03-29)

**rustc version**:
rustc 1.85.0 (4d91de4e4 2025-02-17)

**editor**:
VSCode, Version 1.98.2

This is my first time contributing to a public project, so I appreciate any input for improving this report.

To whom it may concern, I thank you for your time. Hugs and kisses.

**Comment by ChayimFriedman2** [comment]
I'm not sure how we could fix that. If we e.g. try to show "either 8 or 16 bytes", this will become a lot more complicated when there are multiple parameters combined in a struct.

**Comment by TheoAC** [comment]
I don't know what expectations you all have of rust-analyzer's memory layout hover. (Any recommendations on where I could find this?) However, my personal impulse would be to not make any claims about the size of a `&T` at all, unless we already know that `T: Sized`. After all, if I were to write 
```rust
fn foo<T>(t: T) { /* ... */ }
```
then the analyzer *doesn't show any* information about the size of `t` as a parameter for `id`, since it just doesn't have any.
So, along those lines, is it acceptable to just give up on finding/displaying a size for a `&T` when `T: ?Sized`?

This would mean that you only get to see the size of a parameterized struct containing reference types once you know the type behind each reference (or if you know that each is `Sized`). I don't think that that's a problem; it's already the case when generic types are used for fields directly (how big is `struct Thing<T>(T);`?).

All said, I would be interested to know: When do we expect rust-analyzer to provide memory layout information, and what do we expect it to tell us?

**Comment by mozhewen** [comment]
I also encountered this problem when dealing with pointers:
```rust
fn get_size<T: ?Sized>(src: *const T) -> usize {
    size_of_val(&src)
}
```
This function will return 16 (x64 platform) for fat pointers, but hovering on `src` in an editor shows size = 8.
Is the information provided by rust-analyzer just an estimation, or an assurance?

**Comment by ChayimFriedman2** [comment]
The layout information by r-a is generally accurate (we share the same layout implementation with rustc) but here we are cannot provide an accurate value (rustc can't either, only after monomorphization) so we err on the side of providing the more common value. IMO making layout fail over that (not showing layout) will be less useful.

Note that you cannot rely on the layout information in any way, layout of `#[repr(Rust)]` things is not guaranteed.

**Comment by flodiebold** [comment]
IMO if a type parameter is marked `?Sized`, the user probably intends the function to regularly work with unsized `T`, so claiming that `&T` is 8 bytes would confuse more than help. So I'd vote in favor of not showing a size.

**Comment by A4-Tacks** [comment]
IMO, distinguishing is enough, like `size = 8, align = 0x8` and `size = 8*, align = 0x8`
